Bluetooth: Controller: Fix ISO Tx PDU memory pool corruption

cvinayak opened the pull request:
Increase the Controller supported Advertising Data Length
Maximum so that Advertising Data set by Broadcast Audio
Source sample is accepted by the controller.

Also, tune down the supported ISO Tx PDU size to suffice the
Broadcast Audio Source usecase.

Move the removal of BIG context association with the
Periodic Advertising context to when the BIG streams are
release and BIG terminate message is generated.

Add implementation to flush pending ISO Tx PDUs when BIG
events are terminated.

Fix ISO Tx PDU memory pool corruption due to duplicate ISO
Tx PDU buffers repeatedly released into the memory pool.

Fixes #45666.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix removal of BIG association with Periodic Adv

Move the removal of BIG context association with the
Periodic Advertising context to when the BIG streams are
release and BIG terminate message is generated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add implementation to flush Tx on BIG terminate

Add implementation to flush pending ISO Tx PDUs when BIG
events are terminated.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ISO Tx PDU memory pool corruption

Fix ISO Tx PDU memory pool corruption due to duplicate ISO
Tx PDU buffers repeatedly released into the memory pool.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Increase adv data Len for broadcast audio source

Increase the Controller supported Advertising Data Length
Maximum so that Advertising Data set by Broadcast Audio
Source sample is accepted by the controller.

Also, tune down the supported ISO Tx PDU size to suffice the
Broadcast Audio Source usecase.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2414,26 +2414,58 @@ static uint8_t tx_cmplt_get(uint16_t *handle, uint8_t *first, uint8_t last)
+			 *       reference is overloaded to store the Tx
+			 *       fragment count.
+			 *
+			 *       A hack is used here that depends on the fact
+			 *       that memory addresses have a value greater than
+			 *       0xFF, to determined if a node Tx has been
+			 *       released in a prior iteration of this function.
```
So is this a temporary solution? If so, I suggest to create a GH issue to track this and to implement a proper solution.

Thalley replied:
Is implementing this without a "hack" a large task, or what is the reasoning to use a "hack"?

cvinayak replied:
The implementation comes from the original controller contribution, and has not been changed since then.

cvinayak replied:
I will give a thought, this has to done for the ACL Tx ack too where there is difference between ack that is generating a num complete only for data PDU and not for control PDU (inline ENC_REQ PDU) that was transmitted.

mtpr-ot replied:
I have an alternative solution for which I will make a PR today

cvinayak replied:
@mtpr-ot should we merge this PR and thereafter yours? (#49855 LGTM, though you apply the same changes to ACL too).

mtpr-ot replied:
@cvinayak ACL, what do you mean?

mtpr-ot replied:
I think my solution is more intuitive, and if possible, should go in first. You need to re-implement this for current state of main anyway.

cvinayak replied:
> ACL, what do you mean? 

Check this PR changes, ACL connections too use similar bitmask for num complete counting, it will be best if both are fixed.

> I agree, this is very difficult to understand and maintain, so instead of repeating the pattern from node_tx for node_tx_iso, could we perhaps use the available unused bits in node_tx_iso for a state, which can maintain the information instead of overloading the pointer meaning?

#49855 is still the same hack but hiding the code with macros, compared to your original request in this PR was for a redesign.

There seems to be no merge conflicts reported in this PR. As #49855 is just a refactor of the existing hack, it should go after this PR which is a fix for a github issue.

mtpr-ot replied:
@cvinayak Please rebase

Thalley commented:
Running the `broadcast_audio_source` on the 52840DK with this PR:
```
*** Booting Zephyr OS build v3.1.0-rc2-62-g7c895b3d7fd9  ***
Bluetooth initialized
Creating broadcast source
Starting broadcast source
[00:00:00.365,142] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.365,173] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.365,203] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 3.1 Build 0
[00:00:00.365,814] <inf> bt_hci_core: Identity: C4:E7:C8:4B:98:F5 (random)                 
[00:00:00.365,844] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x05f1
[00:00:00.365,844] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff                
Broadcast source started
Waiting 30 seconds before stopped
Sent 1000 total ISO packets
Sent 2000 total ISO packets
Sent 3000 total ISO packets
Stopping broadcast source
Broadcast source stopped
Deleting broadcast source
Broadcast source deleted
Creating broadcast source
Starting broadcast source
[00:00:31.677,581] <err> bt_hci_core: packets count mismatch
[00:00:31.677,642] <err> bt_hci_core: packets count mismatch
Broadcast source started
Waiting 30 seconds before stopped
[00:00:32.914,215] <err> bt_hci_core: No connection for handle 0
[00:00:32.914,306] <err> bt_hci_core: No connection for handle 0
Sent 4000 total ISO packets
Sent 5000 total ISO packets
Sent 6000 total ISO packets
Stopping broadcast source
Broadcast source stopped
Deleting broadcast source
Broadcast source deleted
Creating broadcast source
Starting broadcast source
[00:01:03.005,096] <err> bt_hci_core: packets count mismatch
[00:01:03.005,187] <err> bt_hci_core: packets count mismatch
Broadcast source started
Waiting 30 seconds before stopped
[00:01:04.241,485] <err> bt_hci_core: No connection for handle 0
[00:01:04.241,577] <err> bt_hci_core: No connection for handle 0
```

The same sample running on main (SHA: c534ae294724f3463369363da9818a7a3cf81f95)
```
*** Booting Zephyr OS build v3.1.0-rc2-58-gc534ae294724  ***
Bluetooth initialized
Creating broadcast source
Starting broadcast source
[00:00:00.363,983] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.364,013] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.364,044] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 3.1 Build 0
[00:00:00.364,654] <inf> bt_hci_core: Identity: C4:E7:C8:4B:98:F5 (random)
[00:00:00.364,654] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x05f1
[00:00:00.364,685] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
Broadcast source started
Waiting 30 seconds before stopped
Sent 1000 total ISO packets
Sent 2000 total ISO packets
Sent 3000 total ISO packets
Stopping broadcast source
Broadcast source stopped
Deleting broadcast source
Broadcast source deleted
Creating broadcast source
Starting broadcast source
[00:00:31.675,903] <err> bt_hci_core: packets count mismatch
[00:00:31.675,994] <err> bt_hci_core: packets count mismatch
Broadcast source started
Waiting 30 seconds before stopped
[00:00:32.901,458] <err> bt_hci_core: No connection for handle 0
[00:00:32.901,550] <err> bt_hci_core: No connection for handle 0
Stopping broadcast source
Broadcast source stopped
Deleting broadcast source
Broadcast source deleted
Creating broadcast source
Starting broadcast source
[00:01:02.992,828] <err> bt_hci_core: packets count mismatch
Broadcast source started
Waiting 30 seconds before stopped
[00:01:04.218,170] <err> bt_hci_core: No connection for handle 0
[00:01:04.218,261] <err> bt_hci_core: No connection for handle 0
```

The BUS FAULT mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/45666#issuecomment-1131539301 is no longer present on main (fixed by something else?), and this PR does improve the recreation of the BIG, but the issues with "packets count mismatch" and "No connection for handle 0" are still present.

cvinayak commented:
> The BUS FAULT mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/45666#issuecomment-1131539301 is no longer present on main (fixed by something else?), and this PR does improve the recreation of the BIG, but the issues with "packets count mismatch" and "No connection for handle 0" are still present.

I was able to reproduce the BUS FAULT on upstream/main yesterday. And without this PR, but with #45069 included their will be assertion in `ull_adv_iso.c`. The fault depends on memory layout due to invalid memory released into controller's ISO Tx PDU pool.

Thalley commented:
> > The BUS FAULT mentioned in [#45666 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/45666#issuecomment-1131539301) is no longer present on main (fixed by something else?), and this PR does improve the recreation of the BIG, but the issues with "packets count mismatch" and "No connection for handle 0" are still present.
> 
> I was able to reproduce the BUS FAULT on upstream/main yesterday. And without this PR, but with #45069 included their will be assertion in `ull_adv_iso.c`. The fault depends on memory layout due to invalid memory released into controller's ISO Tx PDU pool.

Can you also reproduce it on `main` today? Is it related to timing, or did something change since yesterday?

cvinayak commented:
![image](https://user-images.githubusercontent.com/6350656/170234637-9a5e2d08-3521-4ce4-b9bd-5a1814acbf00.png)

I am able to reproduce it on the tag: v3.1.0-rc2, with one BT_DBG changed to printk.
And with the commits in this PR cherry-picked the fault is not reproducible.

mtpr-ot committed:
Bluetooth: Controller: Fix ISO TX pool corruption and missing release

- Fix ISO TX data pool corruption due to multiple calls to
  tx_cmplt_get, without removing the nodes
- Added missing release of TX node link in case of ISOAL error

Signed-off-by: Morten Priess <mtpr@oticon.com>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (changes requested):
It looks like the changes in this PR reverts previous commits in ull.c. Please fix.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2464,21 +2464,56 @@ static uint8_t tx_cmplt_get(uint16_t *handle, uint8_t *first, uint8_t last)
+					fragments = (uint32_t)tx_node_iso;
+				}
 
-			if (IS_ADV_ISO_HANDLE(tx->handle)) {
-				/* FIXME: ADV_ISO shall be updated to use ISOAL for
-				 * TX. Until then, assume 1 node equals 1 fragment.
-				 */
-				cmplt += 1;
-			} else {
-				/* We count each SDU fragment completed by this PDU */
-				cmplt += tx_node_iso->sdu_fragments;
```
@mtpr-ot This is not a revert of previous changes in ull.c. Do you mean the comments missing?

cvinayak committed:
Bluetooth: Controller: Use macros in num complt implementation

Use macros to encode the counting of number of completed
packets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2464,21 +2464,56 @@ static uint8_t tx_cmplt_get(uint16_t *handle, uint8_t *first, uint8_t last)
+			if ((((uint32_t)tx_node_iso & ~0xFF) &&
+			     ((p->ll_id == PDU_CIS_LLID_COMPLETE_END) ||
+			      (p->ll_id == PDU_BIS_LLID_COMPLETE_END))) ||
+			    ((uint32_t)tx_node_iso & 0xFF)) {
```
Sorry for not being more precise. You are re-introducing code that was refactored away in a previous commit. This will not work, and will fail our qualification tests. There should be no LLID checks.

mtpr-ot reviewed (approved):
LGTM
